handler/src/main/java/io/netty/handler/ssl/SslContext.java

```diff
@@ -673,6 +673,204 @@ static ApplicationProtocolConfig toApplicationProtocolConfig(Iterable<String> ne
+        private File trustCertChainFile;
+        private TrustManagerFactory trustManagerFactory;
+        private File keyCertChainFile;
+        private File keyFile;
+        private String keyPassword;
+        private KeyManagerFactory keyManagerFactory;
+        private Iterable<String> ciphers;
+        private CipherSuiteFilter cipherFilter = IdentityCipherSuiteFilter.INSTANCE;
+        private ApplicationProtocolConfig apn;
+        private long sessionCacheSize;
+        private long sessionTimeout;
+
+        private Builder(boolean forClient) {
+            this.forClient = forClient;
+        }
+
+        /**
+         * The {@link SslContext} implementation to use. {@code null} uses the default one.
+         */
+        public Builder sslProvider(SslProvider provider) {
+            this.provider = provider;
+            return this;
+        }
+
+        /**
+         * Trusted certificates for verifying the remote endpoint's certificate. The file should
+         * contain an X.509 certificate chain in PEM format. {@code null} uses the system default.
+         */
+        public Builder trustManager(File trustCertChainFile) {
+            this.trustCertChainFile = trustCertChainFile;
+            this.trustManagerFactory = null;
+            return this;
+        }
+
+        /**
+         * Trusted manager for verifying the remote endpoint's certificate. Using a {@code
+         * TrustManagerFactory} is only supported for {@link SslProvider#JDK}; for other providers,
+         * you must use {@link #trustManager(File)}. {@code null} uses the system default.
+         */
+        public Builder trustManager(TrustManagerFactory trustManagerFactory) {
+            this.trustCertChainFile = null;
+            this.trustManagerFactory = trustManagerFactory;
+            return this;
+        }
+
+        /**
+         * Identifying certificate for this host. {@code keyCertChainFile} and {@code keyFile} may
+         * be {@code null} for client contexts, which disables mutual authentication.
+         *
+         * @param keyCertChainFile an X.509 certificate chain file in PEM format
+         * @param keyFile a PKCS#8 private key file in PEM format
+         */
+        public Builder keyManager(File keyCertChainFile, File keyFile) {
+            return keyManager(keyCertChainFile, keyFile, null);
+        }
+
+        /**
+         * Identifying certificate for this host. {@code keyCertChainFile} and {@code keyFile} may
+         * be {@code null} for client contexts, which disables mutual authentication.
+         *
+         * @param keyCertChainFile an X.509 certificate chain file in PEM format
+         * @param keyFile a PKCS#8 private key file in PEM format
+         * @param keyPassword the password of the {@code keyFile}, or {@code null} if it's not
+         *     password-protected
+         */
+        public Builder keyManager(File keyCertChainFile, File keyFile, String keyPassword) {
+            if (!forClient) {
```

Comment:
I'm not sure this justifies duplicating the logic in the builder.  As you said `If calling the method required the proper parameters, even for clients, then once set, there is no way to revert such a configuration`...the builder is just a tool to build objects...the business logic of what is required is in the objects that get constructed.  The builder could be used in transient states while being configured...or repurposed to build something else.  If we update the logic in the underlying classes then we will also have to duplicate the updates in the builder.